app/controllers/measurements_controller.rb

```diff
@@ -2,47 +2,50 @@ class MeasurementsController < ApplicationController
   layout false
 
   def show
-    iteration_time = reference_time - params[:iteration].to_i * interval
-    cache_key = {
-      id: params[:id],
-      day: iteration_time.yday,
-      interval_number: interval_number(iteration_time)
-    }
-
-    @readings = Rails.cache.fetch(cache_key, expires_in: 25.hours) do
-      @measurement = Measurement.find(params[:id])
-      @measurement.
-        readings.
-        includes(:sensor).
-        where('time > ? AND time <= ?', from, to).
-        to_json(include: :sensor)
+    if params[:iteration].blank?
+      last_reading = Reading.where(measurement_id: params[:id]).last
+      readings = readings_json((last_reading&.time || Time.now),
+                               (2 * interval).minutes)
+    else
+      to = reference_time - (params[:iteration].to_i * interval).minutes
+      readings = readings_json(to, 25.hours)
     end
 
-    render json: @readings
+    render json: readings
   end
 
   private
 
+  def readings_json(to, expires_in)
+    cache_key = { id: params[:id], to: to&.to_f }
```

Comment:
Why I will not hit cache again? We are using `to` time set to 0, 15, 30, 45 minutes each hour for iterations and it is repeatable. The same situation is for last reading time (time value is not changed after reading is recorded). I added dedicated tests which check that we are hitting cache, it was also tested manually. Everything seems to be OK.